specification/compute/resource-manager/Microsoft.Compute/DiagnosticRP/preview/2024-03-01-preview/diagnostic.json

```diff
@@ -0,0 +1,695 @@
+      "properties": {
+        "storageAccountId": {
+          "description": "Qualified ID of the storage account.",
+          "type": "string"
+        }
+      },
+      "required": [
+        "storageAccountId"
+      ],
+      "description": "Data used for registering a Storage Account for a Subscription."
+    },
+    "StorageConfigurationResponse": {
+      "type": "object",
+      "properties": {
+        "storageAccountId": {
+          "type": "string",
+          "description": "Fully qualified url for storage account."
+        }
+      },
+      "description": "Api output result when there is an existing storage configuration entry."
+    },
+    "ErrorResponse": {
+      "type": "object",
+      "properties": {
+        "error": {
+          "$ref": "../../../../../../common-types/resource-management/v5/types.json#/definitions/ErrorDetail"
+        }
+      },
+      "description": "An error response from the Compute Diagnostic Resource Provider service."
+    },
+    "ErrorDetail": {
+      "type": "object",
+      "properties": {
+        "code": {
+          "type": "string",
+          "description": "The error code."
+        },
+        "target": {
+          "type": "string",
+          "description": "The target of the particular error."
+        },
+        "message": {
+          "type": "string",
+          "description": "User friendly error message."
+        },
+        "details": {
+          "type": "array",
+          "items": {
+            "$ref": "../../../../../../common-types/resource-management/v5/types.json#/definitions/ErrorDetail"
+          },
+          "description": "The Api error details"
+        },
+        "innerError": {
+          "$ref": "#/definitions/InnerError",
+          "description": "The Api inner error"
+        }
+      },
+      "description": "Error Detail message."
+    },
+    "InnerError": {
+      "type": "object",
+      "properties": {
+        "exceptiontype": {
+          "type": "string",
+          "description": "The exception type."
+        },
+        "errordetail": {
+          "type": "string",
+          "description": "The internal error message or exception dump."
+        }
+      },
+      "description": "Inner error details."
+    },
+    "SpotPlacementRecommenderInput": {
+      "type": "object",
+      "properties": {
+        "desiredLocations": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "The desired regions"
+        },
+        "desiredSizes": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/ResourceSize"
+          },
+          "description": "The desired resource SKUs."
+        },
+        "desiredCount": {
+          "type": "integer",
+          "format": "int32",
+          "description": "Desired instance count per region/zone based on the scope."
+        },
+        "availabilityZones": {
+          "type": "boolean",
```

Comment:
>boolean [](http://example.com/codeflow?start=19&length=7)

API design guideline: 
booleans are not recommended because they're hard to extend in a nonbreaking way
E.g. what if zonal always means 3 zones today, and you want to introduce a 2 zones or 4 zones option in the future?

Reply:
Consider defining a set of string enum values instead.

Reply:
The goal of this API is to return a score for each desired region/sku/zonal input. The availability zones boolean is used to indicate whether the API's scope is on a regional or zonal level. If it's on a zonal level, we will return a score for each available zone in each region as long as the zone exists.


The reason we chose to use a boolean instead of an array or string enum is because the desired regions input is an array, and each region potentially has a different number of zones. Having one string or array to encapsulate all different combination of zones in each region is not sufficient, and in the foreseeable future there's no plan to skip zones in regions if the zone exists.

This API checks which zones exists in each region using Subscription Lifecycle API data. If the number of zones change, the API will return a score which includes new zones for that region, or excludes retired zones.

Reply:
Resolving